### PR TITLE
feat: add caching of constant sub expressions in SympyModels

### LINKED ISSUE #100: Caching of partial calculations
The current Tensorflow code runs relatively slow (about 3x slower than pycompwa), which is related to the fact that tensorwaves is not caching any computations. Even the ones that that stay constant throughout the whole fitting/optimization phase (WignerD functions).

Some flexible and easy to use scheme is necessary to implement local caching.

**Update:** A lot of things have changed since. Now with the sympy + lambdify, the whole model can be optimized much easier. A separate method was added to the Model interface `performance_optimize`.

### PR DESCRIPTION
**Pull request by spflueger**
Closes #100

### CODE REVIEW

**Comment by spflueger** [comment]
I'm keeping it like this for now. Not the best user interface, but we have more issues regarding the Model interface and the gradient as well.

File: src/tensorwaves/model.py [hunk 1]
```diff
@@ -77,6 +143,135 @@ def update_parameters(
...
+                input_args[non_fixed_arg_positions[i]] = x
+
+        def wrapper(*args: Tuple[Any, ...]) -> Any:
+            update_args(*args)
+            return lambdified_model(*input_args)
+
+        return wrapper
+
+    def performance_optimize(self, fix_inputs: DataSample) -> "Model":
+        return NotImplemented
```
Comment (redeboer) [inline src/tensorwaves/model.py:254]: Shouldn't this method return the same as `SympyModel.performance_optimize()`? So returning another `_ConstantSubExpressionSympyModel`?

File: src/tensorwaves/model.py [hunk 2]
```diff
@@ -77,6 +143,135 @@
...
+        if not is_constant:
+            for sub_expr in temp_constant_sub_expression:
+                placeholder = sp.Symbol(f"cached[{str(sub_expr)}]")
+                self.__constant_sub_expressions[placeholder] = sub_expr
+        return is_constant
+
+    def __replace_constant_sub_expressions(
+        self, expression: sp.Expr
+    ) -> sp.Expr:
+        new_expression = copy.deepcopy(expression)
```
Comment (redeboer) [inline src/tensorwaves/model.py:207]: I'm surprised this `deepcopy` is necessary. Does `xreplace`/`subs` modify the expression?

**Review by redeboer — COMMENTED** [review]
Good job!
I left a few comments. It's a bit obscure why the macos job keeps failing now, looking into it. This is a problem that sporadically happened before, so this has to be tackled anyway. (Still, it's strange that this PR makes the macos job more unstable.)

File: src/tensorwaves/model.py (hunk 2 above)
Comment (spflueger) [reply src/tensorwaves/model.py:207]: Good question. I'm actually not sure on the internals of the sympy expression, but the original expression not the be modified

File: src/tensorwaves/model.py (hunk 1 above)
Comment (spflueger) [reply src/tensorwaves/model.py:254]: I want this returned instance to implement the Model Interface

**Review by redeboer — APPROVED** [review]

**Comment by redeboer** [comment]
A series of PRs related to the callbacks is coming up that might fix the CI problem seen here.

**Pull request merged**